Dev-stack-building-thematic-vis (#943)
cambridge-cares/TheWorldAvatar: Deploy/stacks/dynamic/stack-data-uploader/README.md

The list of avaliable options can be found [here][3dcitydb-importer-cli].
The "Database connection options" are set automatically by the `stack-data-uploader` so can be ignored.

#### `"augmentData"`

> This boolean flag controls whether footprint and height would be added to the existing data. It is assumed to be `true` if omitted.

Reviewer: ```suggestion
This boolean flag controls whether footprint and height will be calculated and added to the existing data, if not already present. It is assumed to be `true` if omitted.
```

#### `"discoverThematicSurface"`

If this boolean flag is set to `true`, the uploader will attempt to identify untagged surfaces as roof, wall or ground surfaces. It is assumed to be `false`.

### X building data

The `"XtoCityDB"` data type should be used to load LoD1 (footprint with height) building data that are not CityGML or CityJSON.
The data loader does the following when uploading data: 
1. It uses the GDAL [`ogr2ogr`][ogr2ogr] tool to read in data from a wide variety of file formats and output it to the PostgreSQL database in the stack.
The full list of file formats that `ogr2ogr` supports is given [here][vector-drivers] although some of these might not be available depending on the exact GDAL Docker image being used, see [here][gdal-docker] for details.
2. It uses the 3DCityDB Importer [`impexp import`][3dcitydb-importer] tool to initialise the 3DCityDB schema in the PostgreSQL database in the stack.
3. It uses two SQL scripts to convert the building data uploaded by GDAL and populate the 3DCityDB tables sequentially. The building data will be instantiated as LoD2 buildings with thematic surfaces.
4. Building footprints and heights are added to the uploaded data.

> 5. It writes the processed data in PostgreSQL to a compressed CityGML file using the 3DCityDB Importer [`impexp import`][3dcitydb-importer] tool.

Reviewer: ```suggestion
5. It writes the processed data that has been uploaded to PostgreSQL out to a compressed CityGML file using the 3DCityDB Importer [`impexp import`][3dcitydb-importer] tool.
```

A `"columnMap"` can be specified in the input configuration to inform the stack which columns to look for these information. `"IDval"` is the name of the column that contains the building ID, whereas `"IDname"` is the name of the ID that will be stored in the 3DCityDB tables.

The first SQL script creates two tables from the original table: `"raw_building_XtoCityDB"` and `"raw_surface_XtoCityDB"`. It assumes that the original data is uploaded to the public schema. The `"raw_building_XtoCityDB"` table has the following columns:

1. `"IDval"`: the ID of each building from the original data
2. `"gmlid"`: a randomly generated GML ID of each building
3. `"geom"`: the 3D solid geometry of each building, generated by extruding the `"footprint"` with `"height"` and shifting it with `"elevation"`
5. `"mh"`: the height of each building

The `"raw_surface_XtoCityDB"` table has the following columns:

1. `"building_gmlid"`: the GML ID of the parent building of a surface
2. `"gmlid"`: a randomly generated GML ID of each surface of each building
3. `"class"`: an integer indicating whether a surface is a roof, wall or ground
4. `"geom"`: the 3D polygon geometry of each surface

> In some cases, the original data require more sophisticated processing. User can supply a custom SQL script with `"preprocessSql"` keyword using [File by Value Name](#value-by-file-name). It have to create the `"raw_building_XtoCityDB"` and `"raw_surface_XtoCityDB"` tables in the public schema from the uploaded original data with the column names stated above. In this case, only `"IDname"` and `"IDval"` will take effect and needed to be specified in `"columnMap`.

Reviewer: ```suggestion
In some cases, the original data require more sophisticated processing. Users can supply a custom SQL script with the `"preprocessSql"` keyword using [File by Value Name](#value-by-file-name). The query must create the `"raw_building_XtoCityDB"` and `"raw_surface_XtoCityDB"` tables in the public schema from the uploaded original data with the column names stated above. In this case, only `"IDname"` and `"IDval"` will take effect and needed to be specified in `"columnMap"`.
```
